src/parser.ts

```diff
@@ -415,6 +420,16 @@ export function parseModuleItem(
   line: number,
   column: number
 ): any {
+  if (context & Context.OptionsNext) {
+    const decorators = parseDecorators(parser, context);
```

Comment:
If you aim for performance you should inline this into the `parseDecorators` too avoid the length check. Then you could do `let x = 0` before the loop and `x++`; inside the loop and check `if (x !== 0` at the end of the loop. 
Or even faster  ` if(validdecoratorStart) do iteration consume token  `. If no `validdecoratorStart` the loop will not run, so you automatically have an empty array because that's happen outside the `if statement`.

Reply:
Currently leadingDecorators only apply to module item.

By refactoring parseDecorators, do you mean also refactor parseClassElementList to use `parser.leadingDecorators` too? If so, probably I should just rename it to `parser.decorators`.

Reply:
I didn't think about that. My suggestion was just how to get the best performance.

Reply:
Just tried. Can not use `parser.decorators` for all decorators because of nested `@bar(@foo class B{}) class A {}`.

The check on length here only applies on the syntax error of a module item. I don't think I could push the logic down to parseDecorators.